File: app/src/main/java/com/exa/companydemo/MainActivity.java
package com.exa.companydemo;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.UiModeManager;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;

import android.content.ServiceConnection;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.hardware.SensorManager;
import android.hardware.display.DisplayManager;
import android.hardware.display.VirtualDisplay;
import android.location.Location;
import android.location.LocationManager;
import android.media.Image;
import android.media.projection.MediaProjection;
import android.media.projection.MediaProjectionManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.IBinder;
import android.os.UserHandle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.android.internal.util.ScreenshotHelper;
import com.exa.baselib.BaseConstants;
import com.exa.baselib.base.BaseActivity;
import com.exa.baselib.utils.DateUtil;
import com.exa.baselib.utils.L;

import com.exa.baselib.utils.PermissionUtil;
import com.exa.baselib.utils.ScreenUtils;
import com.exa.baselib.utils.Tools;
import com.exa.baselib.utils.Utils;
import com.exa.companydemo.common.AppInfoActivity;
import com.exa.companydemo.locationtest.LocationActivity;
import com.exa.companydemo.test.BuildTestToast;
import com.exa.companydemo.utils.NetworkManager;
import com.exa.companydemo.widget.EndLineView;
import com.exa.companydemo.widget.Titlebar;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.lang.reflect.Method;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.function.Consumer;

import androidx.core.app.ActivityCompat;

import static android.view.WindowManager.LayoutParams.FLAG_DIM_BEHIND;
import static com.exa.companydemo.TestUtil.isRegisterBroadCast;
import static com.exa.companydemo.TestUtil.mReceiver;
import static com.exa.companydemo.utils.Tools.screenshotView;

/**
 * @author Administrator
 */
public class MainActivity extends BaseActivity {
    private TextView msgT;
    private Button btn1, btn2, btn3, btn4;
    private EndLineView endLineView;
    private EditText editText;
    private boolean isFullScreen;
    private Activity mContext;

    private final String PLATFORM_CONFIG_PATH = "/vendor/etc/data/avm_package_info.json";
    private View toastView;
    private WindowManager windowManager;
    private TextView tv;
    private UiModeManager modeManager;
    private int index = 0;
    private boolean isGome;
    private NetworkManager networkManager;
    private MediaProjectionManager mProjectionManager;

    @Override
    protected int getLayoutId() {
//        ScreenUtils.setFullScreen(this);
//        StatubarUtil.setStatusBarInvasion(this, false);
        return R.layout.activity_main;
    }

    @Override
    protected void initData() {
        L.d("initData");
    }

    @Override
    public void onWindowAttributesChanged(WindowManager.LayoutParams params) {
        super.onWindowAttributesChanged(params);
    }

    @SuppressLint({"ResourceType", "SetTextI18n"})
    @Override
    protected void initView() {
        mContext = this;
        windowManager = App.getContext().getSystemService(WindowManager.class);
        Tools.setScreenBrightness(this, 50);
        modeManager = getSystemService(UiModeManager.class);
        L.d("黑夜模式：" + TestUtil.getUiModeStr(modeManager));
//        checkPermission();
//        TestUtil.registerFullScreenListener(this);
//        TestUtil.registerBroadcast(this);
//        setToolbarId(R.id.toolbar);
        Titlebar toolbar = findViewById(R.id.toolbar);
        toolbar.setNavigationOnClickListener(v -> finish());
        editText = findViewById(R.id.edt);
        msgT = findViewById(R.id.msgT);
        btn4 = findViewById(R.id.btn4);
        endLineView = findViewById(R.id.elv);
        findViewById(R.id.btnApp).setOnClickListener(view -> {
            L.w("点击Location Test");
            startActivity(new Intent(this, LocationActivity.class));
        });
        findViewById(R.id.btnSystemUI).setOnClickListener(view -> {
            L.w("全屏按钮");
            isFullScreen = !isFullScreen;
            if (isFullScreen) {
                ScreenUtils.hideStatusBars(this);
            } else {
                ScreenUtils.showStatusBars(this);
            }
//            startActivity(new Intent(this, SystemUITestActivity.class));
        });
        findViewById(R.id.btnPlay).setOnClickListener(view -> {
            L.w("视频播放");
            startActivity(new Intent(this, VideoPlayerActivity.class));
        });
        findViewById(R.id.btn2).setOnClickListener(view -> {
            L.w("点击跳转  到第二个页面");
            startActivity(new Intent(this, AppInfoActivity.class));
        });
        findViewById(R.id.btn3).setOnClickListener(view -> {
            L.w("进入测试页");
            startActivity(new Intent(this, TestActivity.class));
        });
        findViewById(R.id.btnNightMode).setOnClickListener(v -> {//am start com.android.engmode
            L.w("白天黑夜模式");
            if (modeManager.getNightMode() == UiModeManager.MODE_NIGHT_YES) {
                modeManager.setNightMode(UiModeManager.MODE_NIGHT_NO);
            } else {
                modeManager.setNightMode(UiModeManager.MODE_NIGHT_YES);
            }
            BaseConstants.getHandler().postDelayed(() ->
                    L.w("白天黑夜模式:" + TestUtil.getUiModeStr(modeManager)), 2000);
        });
        findViewById(R.id.btnMode).setOnClickListener(v -> {//am start com.android.engmode
            Utils.openApp(this, "com.android.engmode");
        });
        btn4.setOnClickListener(view -> {
            L.w("测试按钮");
            setText(L.msg);
            try {
                test();
            } catch (Exception e) {
                e.printStackTrace();
                L.e("执行测试异常：" + e.getMessage());
            }
        });
        editText.setOnEditorActionListener((v, actionId, event) -> {
            Tools.hideKeyboard(editText);
            return false;
        });
        // 沉浸式
        ScreenUtils.setStatusBarInvasion(this);
    }

    @SuppressLint({"RestrictedApi", "WrongConstant", "Range"})
    private void test() throws Exception {
        index++;
        L.dd(index);

//        getScreenShotPower();

//        WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams();
//        layoutParams.type = 2019;
//        layoutParams.flags |= 262220;
//        layoutParams.format = -3;
//        layoutParams.width = -1;
//        layoutParams.height = 80;
//        layoutParams.gravity = 80;
//
//        if ((layoutParams.flags & FLAG_DIM_BEHIND) != 0) {
//            L.w("包含FLAG_DIM_BEHIND");
//        } else {
//            L.w("不包含FLAG_DIM_BEHIND");
//        }
//
//        WindowManager.LayoutParams mStateBarParams = new WindowManager.LayoutParams();
//        layoutParams.format = -3;
//        mStateBarParams.type = 2000;
//        mStateBarParams.flags |= -2138832824;
//
//        mStateBarParams.gravity = 53;
//        mStateBarParams.flags &= -33;
//        if ((layoutParams.flags & FLAG_DIM_BEHIND) != 0) {
//            L.w("包含FLAG_DIM_BEHIND");
//        } else {
//            L.w("不包含FLAG_DIM_BEHIND");
//        }

//        new BuildTestToast().addView(mContext);
//        TestUtil.testDialog(this,"ssssss",-1);

//        TestUtil.getSensorData(mContext);
//        L.dd("isTelephonyNetEnable:" + networkManager.isTelephonyDataEnable());

        TestUtil.showAlertDialog(this);
//        TestUtil.testDialog(this,"121212",-1);
//        networkManager.switchTelephonyDataEnable();
//        Toast.makeText(this, "测试Toast: " + index, Toast.LENGTH_SHORT).show();
//        checkPermission();
//        TestUtil.testDialog(this, "ahahh", 0);
//        TestUtil.testFonts(this);
//        TestUtil.showToast(this);
//        startService(new Intent(this, DemoService.class));
//        checkSystemPropertiesReady();

//        Intent intent = new Intent();
//        String pkg = "com.exa.companyclient";
//        String clazz = "com.exa.companyclient.service.MService";
//        intent.setComponent(new ComponentName(pkg,clazz));
//        App.getContext().startService(intent);
    }

    private void printLocation(Location location) {
        setText(location.toString());
        Bundle bundle = location.getExtras();
        if (bundle != null && !bundle.isEmpty()) {
            StringBuilder buffer = new StringBuilder();
            for (String key : bundle.keySet()) {
                buffer.append(key).append("=").append(bundle.get(key).toString()).append(",");
            }
            setText(buffer.toString());
            L.d("LocationBundle=" + bundle + " , " + buffer);
        }
    }

    public static final int REQUEST_MEDIA_PROJECTION = 10001;

    // 申请截屏权限
    private void getScreenShotPower() {
        mProjectionManager = (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
        if (mProjectionManager != null) {
            startActivityForResult(mProjectionManager.createScreenCaptureIntent(), REQUEST_MEDIA_PROJECTION);
        }
    }

    private MediaProjection mMediaProjection;

    // 回调
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_MEDIA_PROJECTION && data != null) {
            if (resultCode == RESULT_OK) {
                MediaProjection mediaProjection = mProjectionManager.getMediaProjection(resultCode, data);
                mediaProjection.registerCallback(new MediaProjection.Callback() {
                    @Override
                    public void onStop() {
                        mediaProjection.unregisterCallback(this);
                        saveScreenshot(mediaProjection);
                    }
                }, null);
                // 截取屏幕内容
                MediaProjection.Callback callback = new MediaProjection.Callback() {
                    @Override
                    public void onStop() {
                        mediaProjection.unregisterCallback(this);
                        saveScreenshot(mediaProjection);
                    }
                };
                mediaProjection.registerCallback(callback, null);
            } else if (resultCode == RESULT_CANCELED) {
                L.d("用户取消");
            }
        }
    }

    private void bindClientService() {
        L.dd();
        Intent intent = new Intent();
        intent.setComponent(new ComponentName("com.exa.companyclient",
                "com.exa.companyclient.service.MService"));
        @SuppressLint("WrongConstant") final boolean isBind = bindService(intent, new ServiceConnection() {
            @Override
            public void onServiceConnected(ComponentName name, IBinder service) {
                L.d("bindService onServiceConnected");
            }

            @Override
            public void onServiceDisconnected(ComponentName name) {
                L.d("bindService onServiceDisconnected");
            }
        }, Context.BIND_AUTO_CREATE | 0x04000000);
        L.e("bind-result:" + isBind);
    }

    private void saveScreenshot(MediaProjection mediaProjection) {
        // 将截取的屏幕内容保存为图片文件
        Bitmap bitmap = Bitmap.createBitmap(2560, 1440, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        Paint paint = new Paint();
        paint.setColor(Color.RED);
        canvas.drawRect(0, 0, 2560, 1440, paint);
        OutputStream outputStream = null;
        try {
            String path = Environment.getExternalStorageDirectory() + "/screenshot.png";
            outputStream = Files.newOutputStream(Paths.get(path));
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream);
            L.d("截屏存储路径：" + path);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (outputStream != null) {
                try {
                    outputStream.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private void screenshot() {
        ImageView imageView = findViewById(R.id.imageView);
        screenshotView(getWindow(), imageView);
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        L.dd(getClass().getSimpleName());
    }

    private void registerReceiver() {
        L.dd("registerTimeUpdateReceiver");
        IntentFilter filter = new IntentFilter();
        filter.addAction("com.gxatek.cockpit.datacenter.action.UPLOAD");
        registerReceiver(mTimeUpdateReceiver, filter);
        setText(DateUtil.getNowDateHM());
    }

    private void registerTimeUpdateReceiver() {
        L.dd("registerTimeUpdateReceiver");
        IntentFilter filter = new IntentFilter();
        filter.addAction(Intent.ACTION_TIME_TICK);
        filter.addAction(Intent.ACTION_TIME_CHANGED);
        registerReceiver(mTimeUpdateReceiver, filter);
        setText(DateUtil.getNowDateHM());
    }

    private final BroadcastReceiver mTimeUpdateReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent == null) return;
            String action = intent.getAction();
            L.d("收到广播：" + action);
            if (action == null || action.isEmpty()) return;
            if (action.equals(Intent.ACTION_TIME_TICK)) {
                //系统每1分钟发送一次广播
                L.d(DateUtil.getNowDateHM());
                setText(DateUtil.getNowDateHM());
            } else if (action.equals(Intent.ACTION_TIME_CHANGED)) {
                //系统手动更改时间发送广播
                L.d(DateUtil.getNowDateHM());
                setText(DateUtil.getNowDateHM());
            }
        }
    };

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
//        L.w("onWindowFocusChanged: hasFocus = " + hasFocus + "\n" + Log.getStackTraceString(new Throwable()));
    }

    @Override
    protected void onResume() {
        super.onResume();
        L.dd();
//        checkPermission();
//        bindScreenSaver();
//        finish();
    }

    @Override
    protected void onStop() {
        super.onStop();
        L.dd(getClass().getSimpleName());
    }

    private void checkPermission() {
        String[] ps = new String[]{Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION};
//        requestPermissions(ps,1);
        PermissionUtil.requestPermission(this, () -> {
            L.d("已授权读写权限");
        }, ps);
    }

    private void setText(String msg) {
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
        String date = format.format(new Date());
        runOnUiThread(() -> {
            String n = date + "\t" + msg + "\n" + msgT.getText().toString();
            msgT.setText(n);
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (isRegisterBroadCast)
            unregisterReceiver(mReceiver);
    }
}

File: app/src/main/java/com/exa/companydemo/MainActivity.kt
package com.exa.companydemo

import android.Manifest
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.app.ActivityOptions
import android.app.UiModeManager
import android.content.*
import android.content.Intent.*
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.net.*
import android.os.*
import android.util.Log
import android.view.*
import android.widget.Button
import android.window.SplashScreen
import androidx.recyclerview.widget.LinearLayoutManager
import com.exa.baselib.BaseConstants
import com.exa.baselib.base.BaseBindActivity
import com.exa.baselib.base.adapter.BaseRecyclerAdapter
import com.exa.baselib.base.adapter.OnClickItemListener
import com.exa.baselib.utils.*
import com.exa.baselib.utils.Tools
import com.exa.companydemo.TestDialog.showDialogFragment
import com.exa.companydemo.TestUtil.*
import com.exa.companydemo.common.AppInfoActivity
import com.exa.companydemo.common.VideoPlayerActivity
import com.exa.companydemo.common.WebActivity
import com.exa.companydemo.databinding.ActivityMainBinding
import com.exa.companydemo.locationtest.LocationActivity
import com.exa.companydemo.toasttest.ToastTestActivity
import com.exa.companydemo.utils.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*


/**
 * @Author lsh
 * @Date 2023/10/23 15:35
 * @Description
 */
class MainActivity : BaseBindActivity<ActivityMainBinding>(), OnClickItemListener {
    private var isFullScreen = false

    private var modeManager: UiModeManager? = null
    private var index = 0
    private var mObjAnim: ObjectAnimator? = null
    private val scope = CoroutineScope(Job() + Dispatchers.IO)
    private lateinit var powerUtil: PowerUtil
    private lateinit var bitmap: Bitmap
    private var mShow = false
    private val handler = Handler(Looper.getMainLooper())

    override fun setContentViewLayoutId(): Int {
//        window.addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN
//                or WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR
//                or WindowManager.LayoutParams.FLAG_SPLIT_TOUCH
//                or WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS
//                or WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION
//                or WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS
//        )
//        window.attributes.fitInsetsTypes = WindowInsets.Type.systemBars()
        return R.layout.activity_main
    }

    override fun initData() {
//        SystemBarUtil.setInvasionStatusBar(this)
//        bind.rv.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
//        window.insetsController?.hide(WindowInsets.Type.systemBars())
    }

    @SuppressLint("ResourceType", "SetTextI18n")
    override fun initView() {
        bind.rv.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        bind.rv.adapter =
            object : BaseRecyclerAdapter<String>(this, R.layout.item_main, btnList, this) {
                override fun onViewHolder(view: View, data: String, position: Int) {
                    view.findViewById<Button>(R.id.btn).text = data
                }
            }
        modeManager = getSystemService(UiModeManager::class.java)
        L.d("黑夜模式：" + getUiModeStr(modeManager))

        val title = (getString(R.string.back) + " (wifi:" + NetworkManager.getInstance(this)
            .getWifiIp() + ":" + SystemProperties.get("service.adb.tcp.port") + ")")
        bind.toolbar.setSubTitle(title)
        bind.toolbar.setNavigationOnClickListener { finish() }
        bind.edit.setOnEditorActionListener { _, _, _ ->
            Tools.hideKeyboard(bind.edit)
            false
        }
        doAfterInitView()
        bind.surfaceView.visibility = View.GONE
        bind.surfaceView.setZOrderOnTop(false)
//        hardKeyTest(this)

//        test()
    }

    private fun doAfterInitView() {
        //设置屏幕亮度
//        Tools.setScreenBrightness(this, 50)
//        checkPermission()
//        test()
        // 加载要显示的图片资源
        bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.win_bg)
//        FordTest.get().test()

//        SystemBarUtil.hideStatusBars(this)
    }

    @SuppressLint(
        "RestrictedApi",
        "WrongConstant",
        "Range",
        "UnspecifiedImmutableFlag",
        "ClickableViewAccessibility", "SetTextI18n"
    )
    @Throws(Exception::class)
    private fun test() {
        App.index++
        val bool = App.index % 2 == 0
        L.dd("${App.index} start------------")

        showDialogFragment(this)
//        startActivity(TestActivity::class.java)

//        moveTaskToBack(true)
//        startService(Intent(this, MDialogService::class.java))
//        FordTest.get().test()
//        if (bool) {
//            TestDialog.getInstance().releaseTimer()
//        } else {
//            TestDialog.getInstance().startPublishTimer()
//        }
//            TestDialog.showMyDialog(this,"12121",2521)
//        showToast(this)

//        TestDialog.showMyDialog(this,"1212",0)
//        TestDialog.showDialogFragment(this)
//        showDialogFragment(this)

//        val packageName = "com.zlingsmart.demo.mtestapp"
//        com.exa.companydemo.utils.Tools.uninstall(this, packageName)
//
//        val deleteObserver = PackageDeleteObserver()
//        val unInstallMethod = packageManager.javaClass.getMethod("deletePackage", String.javaClass, deleteObserver.javaClass, Int.javaClass);
//        unInstallMethod.invoke(packageManager, arrayOf(packageName, deleteObserver, 0))


//        val intent = Intent(this, javaClass)
//        val sender = PendingIntent.getActivity(this, 0, intent, 0)
//        packageManager.packageInstaller.uninstall(packageName,sender)
//            VersionedPackage(
//                packageName,
//                PackageManager.VERSION_CODE_HIGHEST
//            ), sender
//        )

//        mShow = true
//        val isAlive = mInterceptor.asBinder().isBinderAlive
//        L.dd(isAlive)

//        mStartupInterceptor.unregisterHomeKeyInterceptor(mInterceptor)


//        DialogFragmentTest().show(supportFragmentManager, javaClass.name)
//        startShowAnim(bind.image)
//        doSurfaceViewAnimation(this, bind.frame)

//        window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
        L.dd("${App.index} end------------")
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        newConfig.apply {
            L.dd("language=${locale.language} country=${locale.country} uiMode=${uiMode}")
        }
    }

    @SuppressLint("WrongConstant")
    private fun bindClientService() {
        L.dd()
        val intent = Intent()
        intent.component = ComponentName(
            "com.exa.companyclient", "com.exa.companyclient.service.MService"
        )
        val isBind = bindService(intent, object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName, service: IBinder) {
                L.d("bindService onServiceConnected")
            }

            override fun onServiceDisconnected(name: ComponentName) {
                L.d("bindService onServiceDisconnected")
            }
        }, Context.BIND_AUTO_CREATE or 0x04000000)
        L.e("bind-result:$isBind")
    }

    private fun registerReceiver() {
        L.dd("registerTimeUpdateReceiver")
        val filter = IntentFilter()
        filter.addAction("com.gxatek.cockpit.datacenter.action.UPLOAD")
        registerReceiver(mTimeUpdateReceiver, filter)
        setText(DateUtil.getNowDateHM())
    }

    private fun registerTimeUpdateReceiver() {
        L.dd("registerTimeUpdateReceiver")
        val filter = IntentFilter()
        filter.addAction(Intent.ACTION_TIME_TICK)
        filter.addAction(Intent.ACTION_TIME_CHANGED)
        registerReceiver(mTimeUpdateReceiver, filter)
        setText(DateUtil.getNowDateHM())
    }

    private val mTimeUpdateReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            L.d("收到广播：$action")
            if (action.isNullOrEmpty()) return
            if (action == Intent.ACTION_TIME_TICK) {
                //系统每1分钟发送一次广播
                L.d(DateUtil.getNowDateHM())
                setText(DateUtil.getNowDateHM())
            } else if (action == Intent.ACTION_TIME_CHANGED) {
                //系统手动更改时间发送广播
                L.d(DateUtil.getNowDateHM())
                setText(DateUtil.getNowDateHM())
            }
        }
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        L.w("onWindowFocusChanged: hasFocus = $hasFocus")
    }

    private fun checkPermission() {
        val ps = arrayOf(
            Manifest.permission.RECORD_AUDIO,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE
        )
        //        requestPermissions(ps,1);
        PermissionUtil.requestPermission(this, { L.d("已授权") }, ps)
    }

    @SuppressLint("SimpleDateFormat")
    private fun setText(msg: String) {
        val format = SimpleDateFormat(DateUtil.PATTERN_FULL_MS)
        val date = format.format(Date())
        runOnUiThread {
            val n = "$date $msg ${bind.msgT.text}"
            bind.msgT.text = n
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        var msg = javaClass.simpleName
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            msg += " displayId=" + display?.displayId
        }
        L.dd(msg)
    }

    override fun onRestart() {
        super.onRestart()
        val msg = javaClass.simpleName
        L.dd(msg)
    }

    override fun onStart() {
        super.onStart()
    }

    @SuppressLint("SuspiciousIndentation")
    override fun onResume() {
        super.onResume()
        val msg = javaClass.simpleName
        L.dd(msg)
//        checkPermission();
//        bindScreenSaver();
//        finish();
    }

    override fun onPause() {
        super.onPause()
        L.dd(javaClass.simpleName)
    }

    override fun onStop() {
        super.onStop()
        val msg = javaClass.simpleName
        L.dd(msg)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        event?.apply {
            if (keyCode == KeyEvent.KEYCODE_BACK) {
                L.d("Do nothing with KEYCODE_BACK")
//                return true
            }
            if (keyCode == KeyEvent.KEYCODE_HOME) {
                L.d("Do nothing with KEYCODE_HOME")
//                return true
                return true
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onDestroy() {
        super.onDestroy()
        L.dd(javaClass.simpleName)
        if (isRegisterBroadCast) {
            unregisterReceiver(mReceiver)
        }
        App.exit()
    }

    private val btnList = mutableListOf(
        "SystemUI测试",
        "工程模式",
        "原生设置",
        "白天黑夜",
        "Location测试",
        "Toast测试",
        "App列表",
        "视频播放",
        "WebActivity",
        "测试按钮"
    )

    override fun onClickItem(position: Int) {
        L.d("点击按钮:" + btnList[position])
        setText(L.msg)
        when (position) {
            0 -> {
                isFullScreen = !isFullScreen
                if (isFullScreen) {
                    window.decorView.systemUiVisibility =
                        View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    SystemBarUtil.hideStatusBars(this)
//                    SystemBarUtil.hideStatusBar(this)
//                    SystemBarUtil.setInvasionStatusBar(this)
                } else {
//                    SystemBarUtil.setInvasionNone(window)
                    SystemBarUtil.showStatusBars(window)
                }
            }

            1 -> {
                val apps = resources.getStringArray(com.exa.baselib.R.array.engine_mode_pkgs)
                L.dd(Arrays.toString(apps))
                for (item in apps) {
                    if (Utils.openApp(this, item)) {
                        return
                    }
                }
            }

            2 -> {
                val apps = resources.getStringArray(com.exa.baselib.R.array.setting_pkgs)
                for (item in apps) {
                    if (Utils.openApp(this, item)) {
                        return
                    }
                }
            }

            3 -> {
                if (modeManager!!.nightMode == UiModeManager.MODE_NIGHT_YES) {
                    modeManager!!.nightMode = UiModeManager.MODE_NIGHT_NO
                } else {
                    modeManager!!.nightMode = UiModeManager.MODE_NIGHT_YES
                }
                BaseConstants.getHandler().postDelayed({
                    L.w("白天黑夜模式:" + TestUtil.getUiModeStr(modeManager))
                }, 2000)
            }

            4 -> startActivity(LocationActivity::class.java)
            5 -> startActivity(ToastTestActivity::class.java)
            6 -> startActivity(AppInfoActivity::class.java)
            7 -> startActivity(VideoPlayerActivity::class.java)
            8 -> startActivity(WebActivity::class.java)
            9 -> try {
                test()
            } catch (e: Exception) {
                e.printStackTrace()
                Log.e(L.TAG, "执行测试异常：" + e.message, e)
            }

            else -> {}
        }
    }

    override fun onLongClickItem(position: Int) = Unit
}
